Add service to update icloud tracker whenever we want

Bart274 opened the pull request.

Bart274 committed:
Add service to update icloud tracker whenever we want

Bart274 committed:
remove whitespace

Bart274 committed:
if 'name' is defined in config, you can update each individual tracker

Bart274 committed:
if no name is defined, add the username to the service name

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -29,6 +29,7 @@ def setup_scanner(hass, config, see):
     # Get the username and password from the configuration
     username = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
+    name = config.get('name')
```
Please define this as `CONF_NAME` at the top of the file.

balloob replied:
Also, add `username` as second argument so that that is returned when no config entry for name has been found.
```python
name = config.get(CONF_NAME, username)
```

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -78,6 +79,14 @@ def update_icloud(now):
         except PyiCloudNoDevicesException:
             _LOGGER.info('No iCloud Devices found!')
 
+    if name is None:
```
`name` will never be `None` if you follow my other comment so you can simplify this again.

Bart274 replied:
thanks for the comments @balloob , changed my code :)

Bart274 committed:
changed logic, thanks to balloob

Bart274 committed:
devices now update at start of HA, added service to play lost iphone sound

now you can play the lost iphone alert calling the `lost_iphone_<name>` service with your device name in the data as `device_name`

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -44,12 +45,24 @@ def setup_scanner(hass, config, see):
+                             status['name'])
+                device.play_sound()
 
-    track_utc_time_change(hass, keep_alive, second=0)
```
Why did you remove  the keep alive?

Bart274 replied:
the keep alive function isn't necessary anymore, you only get a mail when you login the first time

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -78,6 +90,10 @@ def update_icloud(now):
 
+    hass.services.register('device_tracker',
+                           'update_icloud_' + name, update_icloud)
+    hass.services.call('device_tracker', 'update_icloud_' + name)
```
Don't call services in setup

balloob commented:
It was almost good to go and now you decided to add much more, delaying merging your first feature.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -78,6 +90,10 @@ def update_icloud(now):
         except PyiCloudNoDevicesException:
             _LOGGER.info('No iCloud Devices found!')
 
+    hass.services.register('device_tracker',
```
instead of making a service per device we want 1 service call that uses entity_id to specify which device to target or target all if no entity_id specified.

Bart274 replied:
the service isn't registered per device, but per account. you can't update just one device, you can only update all devices linked to one account

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -44,12 +45,24 @@ def setup_scanner(hass, config, see):
-        _LOGGER.info("Authenticate against iCloud")
+        for device in api.devices:
+            status = device.status()
+            _LOGGER.info('lost_iphone: check device %s', status['name'])
```
This shouldn't be logged

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -44,12 +45,24 @@ def setup_scanner(hass, config, see):
+                device.play_sound()
 
-    track_utc_time_change(hass, keep_alive, second=0)
+    def lost_iphone_service(call):
```
You can merge this with the other lost iPhone method. No need to break in two

Bart274 committed:
change services

now you can call the update_icloud service and have the `accountname` as service data, this must be the username in the config if you haven't used the `name` in the config, otherwise you must use that name. if you don't use `accountname` in the service data, it will update all accounts

the lost_iphone service works the same as before, but now you can also define the `accountname` so you can look for that device in just one account or in all accounts. if you don't put the `devicename` in the service data it will play the alert on all devices for the chosen account

Bart274 committed:
removed logging

Bart274 committed:
rechanged it

rechanged adding one service per account, otherwise the services were only registered for the last account

Bart274 committed:
changed the icloud component so there is only two services for all devices

Now there are only two services for all devices:
- device_tracker.update_icloud which updates the devices given the `accountname` from the service data
- device_tracker.lost_iphone which plays the lost iphone alert on the device from the account given with `accountname` and devicename given with `devicename` in the service data. if there is no `devicename` given, the alert will play on all devices from the account

Bart274 committed:
added description of the icloud services

Bart274 committed:
Update icloud.py

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+ATTR_PASSWORD = 'password'
+ATTR_ACCOUNTNAME = 'accountname'
 
+ICLOUDTRACKERS = {}
```
You can define this inside `setup_scanner`, never used outside.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+    def update_icloud(call):
+        """ Calls the update function of an icloud account """
+        accountname = call.data.get('accountname')
+        entity_id = 'icloud.' + accountname
```
This is not a guarantee to be the exact account name. Instead of adding `'icloud' + ` before all account names, just user `accountname`.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+
+    iclouddevice = Icloud(hass, username, password, name)
+    ICLOUDTRACKERS['icloud.' + name] = iclouddevice
+    # add_devices_callback([iclouddevice])
```
Stale comment.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+ICLOUDTRACKERS = {}
+
+
+# def setup_scanner(hass, config, see, add_devices_callback):
```
Stale comment.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
-        """ Keeps authenticating iCloud connection. """
-        api.authenticate()
-        _LOGGER.info("Authenticate against iCloud")
+class Icloud(Entity):  # pylint: disable=too-many-instance-attributes
```
There is no need to have this extend from an `Entity` as the device tracker does not work with entities. You can just extend from `object`.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+                _LOGGER.exception('Error logging into iCloud Service: %s',
+                                  error)
+
+    @property
```
Remove this. Not being used.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+        return self.api is not None
+
+    @property
+    def state_attributes(self):
```
See, if this was being used, this would have been a **horrible** idea. Publishing authentication info :warning: . Just remove it.

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,49 +20,116 @@
+                                           self.password,
+                                           verify=True)
+            except PyiCloudFailedLoginException as error:
+                _LOGGER.exception('Error logging into iCloud Service: %s',
```
The problem you're having now is that icloud is no longer erroring out when a wrong username or password is specified. Yes you log the error but then go on as if all is fine.

Bart274 committed:
Update icloud.py

balloob commented on homeassistant/components/device_tracker/icloud.py:
```diff
@@ -19,69 +19,126 @@
 
     track_utc_time_change(
-        hass, update_icloud,
+        hass, iclouddevice.update_icloud(see),
```
You need to pass a callback, instead you're calling a method and passing the result. Please test your PRs locally before submitting it.

Bart274 replied:
What do you mean I need to pass a callback? Any example?
I'm running this version in my installation and everything works great, so I am testing it locally

balloob replied:
Icloud devices are no longer polled and I'm pretty sure you're getting
exceptions whenever it tries to poll

On Wed, Feb 17, 2016, 00:31 Bart274 <notifications@github.com> wrote:

> In homeassistant/components/device_tracker/icloud.py
> <https://github.com/balloob/home-assistant/pull/1107#discussion_r53133167>
> :
>
> >
> >      track_utc_time_change(
> > -        hass, update_icloud,
> > +        hass, iclouddevice.update_icloud(see),
>
> What do you mean I need to pass a callback? Any example?
> I'm running this version in my installation and everything works great, so
> I am testing it locally
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1107/files#r53133167>.
>

Bart274 replied:
hmm, the only thing I see in my log around the time it should try to poll is the following error:
```
16-02-17 09:50:00 homeassistant.core: BusHandler:Exception doing job
Traceback (most recent call last):
  File "C:\Python34\lib\site-packages\homeassistant-0.14.0.dev0-py3.4.egg\homeassistant\core.py", line 789, in job_handler
    func(arg)
  File "C:\Python34\lib\site-packages\homeassistant-0.14.0.dev0-py3.4.egg\homeassistant\helpers\event.py", line 186, in pattern_time_change_listener
    action(now)
TypeError: 'NoneType' object is not callable
```
Could this the error you mean?
